---
title: Java 예외처리
date: '2023-06-17'
tags: ['java']
draft: false
---

## 프로그램 오류

프로그램의 오류는 발생시점에 따라서 아래와 같이 분리할 수 있다.

- 컴파일 에러 : 컴파일 할 때 발생하는 에러
- 런타임 에러 : 프로그램의 실행 도중에 발생하는 에러
- 논리적 에러 : 컴파일도 잘되고 실행도 잘되지만 의도한 것과 다르게 동작하는 것

논리적 에러의 한예로는 게임에서 몬스터를 죽였는데도 죽지 않는 경우 같은 것을 말한다.

컴파일 에러와 런타임 에러가 각각 있는 것만 봐도, 컴파일을 성공했다고 해서 런타임 에러가 발생하지 않는 것이 아니라는 걸 알 수 있다. 컴파일러가 실행 도중에 발생 할 수 있는 잠재적인 오류까지 검사할 수가 없기 때문이다.

실행 단계에서의 오류 사항을 방지하기 위해서는 실행 도중 발생할 수 있는 모든 경우의 수를 고려하여 대비하는 것이 필요하다. Java에서는 런타임 단계에서 발생할 수 있는 오류를 `에러`와 `예외` 두가지로 구분하고 있다.

- 에러(error) : 프로그램 코드에 의해서 수습될 수 없는 심각한 오류
- 예외(exception) : 프로그램 코드에 의해서 수습될 수 있는 다소 미약한 오류

에러의 경우는 시스템이 종료되어야 할 수준까지 심각한 상황의 문제여서 개발자가 미리 예측하여 방지할 수 없는 것을 의미한다.

반면에 예외는 개발자가 구현한 로직에서 발생한 실수나 사용자의 영향에 의해 발생하는 것을 의미한다. 개발자가 구현한 로직에서 발생하는 것이기 때문에, 이를 미리 예측하여 방지하는 작업을 할 수 있는데 이것을 바로 예외처리라고 한다.

## Exception class

모든 예외의 최고 조상 class는 `Exception`이고 상속 계층도는 아래와 같다.

<img
  width="440"
  alt="image"
  src="https://github.com/Jiyun-Parkk/algorithm-test/assets/72537762/45dac525-95b0-4147-bf43-da0104e1192b"
/>

위 그림과 같이 예외는 두 그룹으로 나눠질 수 있다.

- Exception 클래스와 그 자손들
- RuntimeException과 그 자손들

## Exception과 RuntimeException

- Exception 클래스와 그 자손들: 사용자의 실수와 같은 외적인 요인에 의해 발생하는 예외
- RuntimeException 클래스와 그 자손들: 프로그래머의 실수로 발생하는 예외

Exception 클래스의 자식들은 사용자의 의해 발생하는 것들인데, 예를 들어서 타입과 맞지 않는 데이터를 입력했다던가, 존재하지 않은 파일의 이름을 입력했다던가 등이 있다.

> 💡 checked exception: Exception 클래스의 직계 자식들을 checked exception이라고도 하는데, 이는 반드시 에러처리를 해야하는 특징이 있다.(try-catch활용)

RuntimeException 클래스들은 자바의 프로그래밍 요소들과 관계가 깊은데, 예를 들면 배열의 범위를 벗어났다거나, 값이 Null인 참조변수의 멤버를 호출하려 했다던가, 클래스간의 형변환을 잘못했다거나, 정수를 0으로 나누려고 한 경우 등등이다.

따라서 RuntimeException이 발생하면 나부터 의심을 해보아야 한다.

> 💡 unchecked exception: RuntimeException의 직계 자식들은 unchecked exception인데, 이는 checked exception과 달리 에러처리를 강제하지 않는다.

## 예외처리하기

예외처리란 프로그램 실행 시 발생할 수 있는 예기치 못한 예외의 발생에 대비한 코드를 미리 작성하는 것을 말한다. 여기서 중요한 것은 예외처리의 목적이다.

예외처리의 목적은, 예외로 인한 비정상적인 프로그램의 종료를 막고 정상적인 실행상태를 유지할 수 있도록 하기 위함이다.

예외를 처리하기 위해서는 `try-catch`구문을 활용한다. `try-catch` 구문에서 예외가 발생한 경우와 발생하지 않았을 때 흐름이 달라진다.

```text
1. try 블럭 내에서 예외가 발생한 경우
- 발생한 예외와 일치하는 catch블럭이 있는지 확인
- 일치하는 catch 블럭을 찾게 되면 해당 catch 블럭의 코드를 실행
- 전체 try-catch 구문을 빠져나가서 다음 문장을 계속해서 수행한다.
- 만약 일치하는 catch 블럭을 찾지 못하면, 예외는 처리되지 않는다.
2. try 블럭 내에서 예외가 발생하지 않은 경우
- catch 블럭을 거치지 않고 전체 try-catch문을 빠져나가서 수행을 하게된다.
```

try 블럭 내에 있는 코드에서 예외가 발생하면 위 과정과 같이 해당하는 예외를 찾기위해 catch 구문을 순차적으로 검사하는데, 발생한 예외 클래스의 인스턴스에 instanceof 연산자를 통해 검사를 하게 된다. 검사결과가 true인 catch 블럭을 만날때까지 검사를 계속 하게 된다.

참고로 Exception 클래스는 모든 예외의 부모이기 때문에, Exception 하나만 걸어놓으면 어떤 종류의 예외가 발생해도 catch 블럭에 의해 처리가 된다.

### 멀티 catch 블럭

catch의 괄호`()`안에 예외처리를 할 예외의 종류를 넣어주는데, `|`이 기호로 여러개의 예외를 한번에 넣을 수 있다. 이를 멀티 catch 블럭이라고 하는데, 추가할 수 있는 개수는 제한이 없다. 한가지 유념해야할 점은 `||`의 or 연산자가 아니다.

그런데 만약! 멀티 캐치로 연결된 예외 클래스가 조상과 자손의 관계라면 컴파일 에러가 발생한다.

왜냐하면 앞서서 Exception 클래스로만 예외처리를 하면 모든 예외가 처리되었던 것처럼, 조상 클래스가 있다면 자식 클래스는 자동으로 처리가 되기 때문이다.

## printStackTrace()와 getMessage()

- printStackTrace() : 호출스택에 대한 정보와 예외 메세지
- getMessage() : 발생한 예외클래스의 인스턴스에 저장된 메세지를 얻을 수 있다

## 메서드에 예외 선언하기

메서드에 예외를 선언하는 것은 예외처리 방법 중 하나로, 메서드 선언부에 `throw`키워드를 활용해서 예외를 적어주기만 하면 된다.

```java
public void main(String[] args) throws IOException {
    //..
}
```

위와 같이 작성하면 `IOException`이 발생할 가능성이 있다는 뜻이다. 기존의 많은 언어들에서는 메서드에 예외선언을 하지 않기 때문에, 경험이 많은 프로그래머가 아니고서 어떤 상항에서 어떤 예외가 발생할 가능성이 있는지 예측하기가 힘들었기 때문에 대비하는 것이 어려웠다고 한다.

하지만 자바에서는 메서드를 작성할때 메서드 내에서 발생할 가능성이 있는 예외를 명시해주기 때문에 개발자의 짐을 한층 덜어주는 역할을 한다.

## 커스텀 예외 만들기

기존에 정의된 예외 클래스 외에도 필요에 따라서 프로그래머가 새로운 예외 클래스를 정의해서 사용할 수 있다.

```java
class MyException extends Exception {
    private final int ERR_CODE;
    MyException(String msg, int errCode) {
        super(msg); // 조상인 Exception 클래스의 생성자를 호출한다.
        ERR_CODE = errCode;
    }
    MyException(String msg) {
        this(msg, 100);
    }
    public int getErrCode() {
        return ERR_CODE; // getMessage()와 함께 사용될 것이다.
    }
}
```

기존에는 주로 Exception을 상속받아서 `checked exception`으로 작성하는 경우가 많았지만 요즘에는 선택적으로 처리 할 수 있도록 RuntimeException으로 작성하는 쪽으로 바뀌고 있다고 한다. 왜냐하면 `checked exception`은 무조건 try-catch 처리를 해주어야 하기 때문에 자칫 코드가 복잡해질 가능성이 있기 때문이다.

[참고 - 자바의 정성 예외처리]

[오류와 예외](https://toneyparky.tistory.com/40)

[Exception handling in java](https://www.javatpoint.com/exception-handling-in-java)

[checked exception & unchecked exception](https://devlog-wjdrbs96.tistory.com/351)
